testing/sdk_package_registry.py

```diff
@@ -0,0 +1,243 @@
+    for file_path, name, version in dcos_files_list:
+        rc, out, err = sdk_cmd.run_raw_cli(' '.join([
+            'registry',
+            'add',
+            '--dcos-file={}'.format(file_path),
+            '--json'
+        ]))
+        assert rc == 0
+        assert len(json.loads(out)['packages']) > 0, 'No packages were added'
+        wait_until_cli_condition(
+            ' '.join([
+                'registry',
+                'describe',
+                '--package-name=' + name,
+                '--package-version=' + version,
+                '--json'
+            ]),
+            lambda code, out, err: code == 0 and
+            json.loads(out).get('status') == 'Added'
+        )
+
+
+@retrying.retry(
+    stop_max_delay=5 * 60 * 1000,
+    wait_fixed=5 * 1000
+)
+def wait_until_cli_condition(
+        cmd: str,
+        check: Callable[[int, str, str], bool]
+) -> None:
+    code, stdout, stderr = sdk_cmd.run_raw_cli(cmd)
+    assert check(code, stdout, stderr)
+
+
+def build_dcos_files_from_stubs(
+        stub_universe_urls: List,
+        dcos_files_path: str,
+        tmpdir_factory  # _pytest.TempdirFactory
+) -> List[Tuple[str, str, str]]:
+    if not len(stub_universe_urls):
+        return stub_universe_urls
+    package_file_paths = []
+    for repo_url in stub_universe_urls:
+        headers = {
+            "User-Agent": "dcos/{}".format(sdk_utils.dcos_version()),
+            "Accept": "application/vnd.dcos.universe.repo+json;"
+                      "charset=utf-8;version={}".format('v4'),
+        }
+        req = urllib.request.Request(repo_url, headers=headers)
+        with urllib.request.urlopen(req) as f:
+            data = json.loads(f.read().decode())
+            for package in data['packages']:
+                package_file_paths.append(
+                    build_dcos_file_from_universe_definition(
+                        package,
+                        dcos_files_path,
+                        tmpdir_factory
+                    )
+                )
+    return package_file_paths
+
+
+def build_dcos_file_from_universe_definition(
+        package: Dict,
+        dcos_files_path: str,
+        tmpdir_factory  # _pytest.TempdirFactory
+) -> Tuple[str, str, str]:
+    """
+    Build the .dcos file if its not already present in the given directory.
+    Returns a Tuple containing (path of .dcos file, name, and version)
+    """
+    # TODO Ideally we should `migrate` and then `build`.
+    name = package['name']
+    version = package['version']
+    target = os.path.join(dcos_files_path, '{}-{}.dcos'.format(name, version))
+    if os.path.isfile(target):
+        log.info('Skipping build, using cached file : {}'.format(target))
+    else:
+        del package['releaseVersion']
+        del package['selected']
+        package_json_file = tmpdir_factory\
+            .mktemp(sdk_utils.random_string())\
+            .join(sdk_utils.random_string())
+        package_json_file.write(json.dumps(package))
```

Comment:
- `dumps` seemed more cleaner way to do it as i need to get the abs path of the file as a `str` to pass to the registry command and i don't want more nesting.
- Passing named temporary file may not be what we want because we have a case where if we already have a `.dcos` file there is no work to be done and we don't need any temporary files. so passing `tmpdir_factory` makes sense (as this lazily creates files/directories as needed).

Reply:
Thanks for the clarification.